src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+                        matched_offsets.insert(partition.clone(), (*timestamp, *offset));
+                    assert_eq!(old_value, None);
+                } else {
+                    assert_eq!(bindings.len(), 1);
+                    // Unable to match any more from this partition
+                    break;
+                }
+            }
+
+            // TODO(#12267): Properly downgrade `since` of the timestamp collection based on the source data collection.
+
+            // XXX: should clamp to round timestamp_frequency??
+            let new_ts = (self.now)();
+            let new_upper = Antichain::from_elem(new_ts + 1);
+
+            let new_bindings: Vec<_> = observed_max_offsets
+                .iter()
+                .filter(|(partition, _offset)| !matched_offsets.contains_key(partition))
+                .map(|(partition, new_max_offset)| {
+                    // Attempt to commit up to max offsets at current timestamp
+                    let current_offset = self
+                        .persisted_timestamp_bindings
+                        .entry(partition.clone())
+                        .or_insert_with(VecDeque::new)
+                        .back()
+                        .map(|(_ts, offset)| *offset)
+                        .unwrap_or_default();
+                    let diff = *new_max_offset - current_offset;
+                    assert!(
+                        diff > 0,
+                        "Diff previously validated to be positive: {:?}",
+                        diff
+                    );
+                    (partition, diff)
+                })
+                .collect();
+
+            if new_bindings.is_empty() {
+                assert_eq!(matched_offsets.len(), observed_max_offsets.len());
+                // This should be a sensible value for the source to downgrade its capability to after emitting the messages that we return
+                let progress = matched_offsets
+                    .values()
+                    .map(|(ts, _offset)| *ts)
+                    .chain(std::iter::once(
+                        *self.read_progress.elements().first().unwrap_or(&u64::MAX),
+                    ))
+                    .min()
+                    .unwrap_or_else(Timestamp::minimum);
+                return Ok(Some((matched_offsets, Antichain::from_elem(progress))));
+            }
+
+            let compare_and_append_result = self
+                .write_handle
+                .compare_and_append(
+                    new_bindings
+                        .iter()
+                        .map(|(partition, diff)| ((&(), *partition), &new_ts, diff)),
+                    self.write_upper.clone(),
+                    new_upper.clone(),
+                )
+                .await
+                .expect("Timestamper CAA")
+                .expect("Timestamper CAA 2");
+
+            self.write_upper = match compare_and_append_result {
+                Ok(()) => new_upper,
```

Comment:
Shouldn't there be some logic that leads to the function returning in this branch since everything went as planned?

Reply:
as currently written, we'd have to merge `matched_offsets` and `new_bindings` and return _that_, also calculating the appropriate `progress` return.  I'm about to change the computation of the `progress` to be much more complex (because I believe it's necessary to do so! The current implementation often downgrades too soon).

Because both of those (and especially `progress`) are quite subtle, I'd prefer to have a single place that computes them and returns them.  And I don't currently believe that the one single place should be here because we want to clearly handle the case where we don't actually need to commit new bindings (becasue they've been previously written by another persist client)

Though, note that with your other feedback, squashing outstanding bugs, I may end up moving things around